backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java

```diff
@@ -734,32 +750,77 @@ private String getRootName() {
+        }
+      }
+      return HashMap.ofAll(result);
+    }
+
+    IGitMacheteRepositorySnapshot discoverGitMacheteRepository(int mostRecentlyCheckedOutBranchesCount)
+        throws GitCoreException, GitMacheteException {
+
+      List<String> localBranchNames = localBranches.map(lb -> lb.getName());
+      var branchNamesFixedRootPartition = localBranchNames.partition(Predicate.isEqual("master"));
+      List<String> fixedRootBranchNames = branchNamesFixedRootPartition._1;
+      List<String> nonFixedRootBranchNames = branchNamesFixedRootPartition._2;
+      List<String> freshNonFixedRootBranchNames;
+
+      // Let's only leave at most the given number of most recently checked out ("fresh") branches.
+      if (nonFixedRootBranchNames.size() <= mostRecentlyCheckedOutBranchesCount) {
+        freshNonFixedRootBranchNames = nonFixedRootBranchNames;
+      } else {
+        Map<String, Instant> lastCheckoutTimestampByBranchName = deriveLastCheckoutTimestampByBranchName();
+
+        var freshAndStaleNonFixedRootBranchNames = nonFixedRootBranchNames
+            .sortBy(branchName -> lastCheckoutTimestampByBranchName.getOrElse(branchName, Instant.MIN))
+            .reverse()
+            .splitAt(mostRecentlyCheckedOutBranchesCount);
+        freshNonFixedRootBranchNames = freshAndStaleNonFixedRootBranchNames._1;
+
+        LOG.debug(() -> "Skipping stale branches from the discovered layout: "
+            + freshAndStaleNonFixedRootBranchNames._2.mkString(", "));
+      }
+
+      // Let's use SortedMaps to ensure a deterministic result.
+      SortedMap<String, MyBranchLayoutEntry> entryByFixedRootBranchNames = fixedRootBranchNames
           .toSortedMap(name -> Tuple.of(name, new MyBranchLayoutEntry(name)));
-      List<IBranchLayoutEntry> roots = List.empty();
+      SortedMap<String, MyBranchLayoutEntry> entryByFreshNonFixedRootBranch = freshNonFixedRootBranchNames
+          .toSortedMap(name -> Tuple.of(name, new MyBranchLayoutEntry(name)));
+      SortedMap<String, MyBranchLayoutEntry> entryByIncludedBranchName = entryByFixedRootBranchNames
+          .merge(entryByFreshNonFixedRootBranch);
+      LOG.debug(() -> "Branches included in the discovered layout: " + entryByIncludedBranchName.keySet().mkString(", "));
+
+      // `roots` may be an empty list in the rare case there's no `master` branch in the repository.
+      List<MyBranchLayoutEntry> roots = entryByFixedRootBranchNames.values().toList();
 
-      for (var branchEntry : entryByBranchName.values()) {
-        Seq<String> upstreamCandidates = entryByBranchName.values()
+      // Skipping the upstream inference for fixed roots (currently just `master`) and for the stale non-fixed-root branches.
+      for (var branchEntry : entryByFreshNonFixedRootBranch.values()) {
+        // Note that stale non-fixed-root branches are never considered as candidates for an upstream.
+        Seq<String> upstreamCandidateNames = entryByIncludedBranchName.values()
```

Comment:
While we are here (and in other places): shouldn't we use `parent`/`child` naming convention according to PR #411?

Reply:
Whoops... actually here both `parent` (branch layout naming) and `upstream` (git machete naming) are kinda justified... but now I see some places in this file where `parent` is used for git machete stuff... fixing on a follow-up PR

Reply:
Done on #428